pkg/controller/batchrelease/batchrelease_special_cases_handler.go

```diff
@@ -17,215 +17,253 @@ limitations under the License.
 package batchrelease
 
 import (
-	"time"
+	"reflect"
 
 	"github.com/openkruise/rollouts/api/v1alpha1"
 	"github.com/openkruise/rollouts/pkg/controller/batchrelease/workloads"
 	"github.com/openkruise/rollouts/pkg/util"
 	v1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/klog/v2"
-	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
 const (
-	Keep        = "Keep"
-	Abort       = "Abort"
-	Start       = "Start"
-	Restart     = "Restart"
-	Finalize    = "Finalize"
-	Terminating = "Terminating"
-	Recalculate = "Recalculate"
+	Keep         = "Keep"
+	Start        = "Start"
+	Restart      = "Restart"
+	Finalize     = "Finalize"
+	Terminating  = "Terminating"
+	Recalculate  = "Recalculate"
+	Reinitialize = "Reinitialize"
 )
 
-func (r *Executor) checkHealthyBeforeExecution(controller workloads.WorkloadController) (needStopThisRound bool, result reconcile.Result) {
+func (r *Executor) checkHealthBeforeExecution(controller workloads.WorkloadController) (needStopThisRound bool, result reconcile.Result) {
 	var reason string
 	var message string
-	var action string
 	var needRetry bool
+	var stateTrans string
 
 	// sync the workload info and watch the workload change event
 	workloadEvent, workloadInfo, err := controller.SyncWorkloadInfo()
 
-	// Note: must keep the order of the following cases
+	// Handle the special cases about the workload, include:
+	//   (1). Get workload info err
+	//   (2). Workload was deleted
+	//   (3). Workload is created
+	//   (4). Workload scale when rollout
+	//   (5). Workload rollback when rollout
+	//   (6). Workload revision changed when rollout
+	//   (7). Workload is at unexpected/unhealthy state
+	//   (8). Workload is at unstable state, its workload info is untrustworthy
 	switch {
-	case r.releasePlanTerminating():
-		reason = "PlanTerminating"
-		message = "release plan is terminating, terminate the release plan"
-		action = Terminating
-
-	case r.workloadHasGone(err):
-		reason = "WorkloadGone"
-		message = "target workload has gone, cancel the release plan"
-		action = Terminating
-
-	case client.IgnoreNotFound(err) != nil:
+	// handle the case of IgnoreNotFound(err) != nil
+	case isGetWorkloadInfoError(err):
 		reason = "GetWorkloadError"
 		message = err.Error()
 		needRetry = true
-		action = Keep
+		stateTrans = Keep
```

Comment:
can we call signalXXX function directly, so we don't have to understand the meaning the puzzling state name